app/http.php

```diff
@@ -0,0 +1,1039 @@
+                // Try to remove with contaier name instead of ID
+                $orchestration->remove($runtimeId, true);
+            } catch (Throwable $th) {
+                // If fails, means initialization also failed.
+                // Contianer is not there, no need to remove
+            }
+        }
+
+        // Release orchestration back to pool, we are done with it
+        $orchestrationPool->push($connection);
+
+        $response
+            ->setStatusCode(Response::STATUS_CODE_CREATED)
+            ->json($container);
+    });
+
+
+App::get('/v1/runtimes')
+    ->desc("List currently active runtimes")
+    ->inject('activeRuntimes')
+    ->inject('response')
+    ->action(function (Table $activeRuntimes, Response $response) {
+        $runtimes = [];
+
+        foreach ($activeRuntimes as $runtime) {
+            $runtimes[] = $runtime;
+        }
+
+        $response
+            ->setStatusCode(Response::STATUS_CODE_OK)
+            ->json($runtimes);
+    });
+
+App::get('/v1/runtimes/:runtimeId')
+    ->desc("Get a runtime by its ID")
+    ->param('runtimeId', '', new Text(64), 'Runtime unique ID.')
+    ->inject('activeRuntimes')
+    ->inject('response')
+    ->action(function (string $runtimeId, Table $activeRuntimes, Response $response) {
+        $activeRuntimeId = $runtimeId; // Used with Swoole table (key)
+
+        if (!$activeRuntimes->exists($activeRuntimeId)) {
+            throw new Exception('Runtime not found', 404);
+        }
+
+        $runtime = $activeRuntimes->get($activeRuntimeId);
+
+        $response
+            ->setStatusCode(Response::STATUS_CODE_OK)
+            ->json($runtime);
+    });
+
+App::delete('/v1/runtimes/:runtimeId')
+    ->desc('Delete a runtime')
+    ->param('runtimeId', '', new Text(64), 'Runtime unique ID.', false)
+    ->inject('orchestrationPool')
+    ->inject('activeRuntimes')
+    ->inject('response')
+    ->action(function (string $runtimeId, Pool $orchestrationPool, Table $activeRuntimes, Response $response) {
+        $activeRuntimeId = $runtimeId; // Used with Swoole table (key)
+        $runtimeId = System::getHostname() . '-' . $runtimeId; // Used in Docker (name)
+
+        if (!$activeRuntimes->exists($activeRuntimeId)) {
+            throw new Exception('Runtime not found', 404);
+        }
+
+        Console::info('Deleting runtime: ' . $runtimeId);
+
+        try {
+            $connection = $orchestrationPool->pop();
+            $orchestration = $connection->getResource();
+            $orchestration->remove($runtimeId, true);
+            $activeRuntimes->del($activeRuntimeId);
+            Console::success('Removed runtime container: ' . $runtimeId);
+        } finally {
+            isset($connection) && $orchestrationPool->push($connection);
+        }
+
+        $response
+            ->setStatusCode(Response::STATUS_CODE_OK)
+            ->send();
+    });
+
+App::post('/v1/execution')
+    ->desc('Create an execution')
+    // Execution-related
```

Comment:
Again, it will become very important to decouple the execution from runtime creation. I know we discussed this before, but a few new concerns have shown up, mainly around a lot more params and metadata from Appwrite that will be required for creating a runtime which will make this endpoint very complex.

Reply:
If we remove this from there, the process of executing function will need to be:

1. Try to execute function
2. If error code X, try to create runtime
3. Try to execute function again

Not to mention retries that will need to be implemented. Not only this makes the process slower, it also makes it impossible to use without some kind of SDK.. Normal HTTP communication will get extremely noisy.

Reply:
I am fully aware. It will make the first cold start a few ms slower, but would provide a better design. We can still have some retry logic inside the runtime creation endpoint. An SDK will probably be needed, technically we already have one in Appwrite - we need to try and keep very simple. The first execute and on failure create new runtime, should be Appwrite logic rather than the SDK.

Reply:
Keeping it as is for now.